src/state/config.py

```diff
@@ -0,0 +1,77 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""gateway-api-integrator configuration."""
+
+import dataclasses
+import itertools
+import typing
+
+import ops
+from pydantic import BaseModel, Field, ValidationError
+
+
+class InvalidCharmConfigError(Exception):
+    """Exception raised when a charm configuration is found to be invalid.
+
+    Attrs:
+        msg (str): Explanation of the error.
+    """
+
+    def __init__(self, msg: str):
+        """Initialize a new instance of the InvalidCharmConfigError exception.
+
+        Args:
+            msg (str): Explanation of the error.
+        """
+        self.msg = msg
+
+
+@dataclasses.dataclass(frozen=True)
```

Comment:
`class CharmConfig(BaseModel, frozen=True)` can be used too.

In the case of going through dataclasses, another option is to use pydantic dataclasses (see https://docs.pydantic.dev/latest/concepts/dataclasses/)

Reply:
thanks! I switched to pydantic dataclass as validation was not being done properly